src/Microsoft.VisualStudio.Threading/JoinableTask.cs

```diff
@@ -297,20 +297,29 @@ internal SynchronizationContext ApplicableJobSyncContext
         {
             get
             {
-                using (this.Factory.Context.NoMessagePumpSynchronizationContext.Apply())
+                if (this.Factory.Context.IsOnMainThread)
                 {
-                    lock (this.owner.Context.SyncContextLock)
+                    if (this.mainThreadJobSyncContext == null)
                     {
-                        if (this.Factory.Context.IsOnMainThread)
+                        using (this.Factory.Context.NoMessagePumpSynchronizationContext.Apply())
                         {
-                            if (this.mainThreadJobSyncContext == null)
+                            lock (this.owner.Context.SyncContextLock)
                             {
-                                this.mainThreadJobSyncContext = new JoinableTaskSynchronizationContext(this, true);
+                                if (this.mainThreadJobSyncContext == null)
+                                {
+                                    this.mainThreadJobSyncContext = new JoinableTaskSynchronizationContext(this, true);
+                                }
                             }
-
-                            return this.mainThreadJobSyncContext;
                         }
-                        else
+                    }
+
+                    return this.mainThreadJobSyncContext;
+                }
+                else
+                {
+                    using (this.Factory.Context.NoMessagePumpSynchronizationContext.Apply())
+                    {
```

Comment:
💡  It also appears that `SynchronouslyBlockingThreadPool` latches in the `true` state, so if the value is `true` and `threadPoolJobSyncContext` already has a value, then the lock can be safely avoided.

Reply:
`SynchronouslyBlockingThreadPool` actually latches in the `false` state. It's either always false, or it starts `true` and transitions to `false` later.